csaf-validation/src/main/kotlin/io/github/csaf/sbom/validation/tests/Tests.kt

```diff
@@ -0,0 +1,289 @@
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are not defined: ${notDefined.joinToString(",")}")
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.5](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#615-multiple-definition-of-product-group-id)
+ */
+object Test615MultipleDefinitionOfProductGroupID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val definitions = doc.gatherProductGroups().map { it.group_id }
+
+        val duplicates = definitions.duplicates()
+
+        return if (duplicates.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are duplicate: ${duplicates.keys.joinToString(",")}")
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.6](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#616-contradicting-product-status).
+ */
+object Test616ContradictingProductStatus : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val affected = mutableSetOf<String>()
+        val notAffected = mutableSetOf<String>()
+        val fixed = mutableSetOf<String>()
+        val underInvestigation = mutableSetOf<String>()
+
+        val contradicted = mutableSetOf<String>()
+
+        for (vulnerability in doc.vulnerabilities ?: listOf()) {
+            affected.clear()
+            notAffected.clear()
+            fixed.clear()
+            underInvestigation.clear()
+
+            vulnerability.product_status.gatherAffectedProductReferencesTo(affected)
+            vulnerability.product_status.gatherNotAffectedProductReferencesTo(notAffected)
+            vulnerability.product_status.gatherFixedProductReferencesTo(fixed)
+            vulnerability.product_status.gatherUnderInvestigationProductReferencesTo(
+                underInvestigation
+            )
+
+            contradicted += affected.intersect(notAffected)
+            contradicted += affected.intersect(fixed)
+            contradicted += affected.intersect(underInvestigation)
+            contradicted += notAffected.intersect(fixed)
+            contradicted += notAffected.intersect(underInvestigation)
+            contradicted += fixed.intersect(underInvestigation)
+        }
+
+        return if (contradicted.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf(
+                    "The following IDs have contradicting statuses: ${contradicted.joinToString(",")}"
+                )
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.7](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#617-multiple-scores-with-same-version-per-product).
+ */
+object Test617MultipleScoresWithSameVersionPerProduct : Test {
+    override fun test(doc: Csaf): ValidationResult {
```

Comment:
I *think* this test is not entirely in-line with the specification since it checks whether the CVSS-version of the vulnerability item of a product is unique. However, the spec talks about CVSS-Vectors of the same version.

In theory, this could be interpreted that we could duplicate the entire entry (i.e., have multiple identical CVSS-Vectors for the same product). Maybe, this could be clarified in the specification

Reply:
Not sure if I follow 100 %, Aren't you referring to this state here: https://github.com/oasis-tcs/csaf/blob/master/csaf_2.0/test/validator/data/mandatory/oasis_csaf_tc-csaf_2_0-2021-6-1-07-01.json. And this test vector we fail "correctly".

Reply:
No, I was wondering how it's supposed to behave in such a case (modified the resource from your link a bit):
```json
{
  "document": {
    "category": "csaf_base",
    "csaf_version": "2.0",
    "publisher": {
      "category": "other",
      "name": "OASIS CSAF TC",
      "namespace": "https://csaf.io"
    },
    "title": "Mandatory test: Multiple Scores with same Version per Product (failing example 1)",
    "tracking": {
      "current_release_date": "2021-07-21T10:00:00.000Z",
      "id": "OASIS_CSAF_TC-CSAF_2.0-2021-6-1-07-01",
      "initial_release_date": "2021-07-21T10:00:00.000Z",
      "revision_history": [
        {
          "date": "2021-07-21T10:00:00.000Z",
          "number": "1",
          "summary": "Initial version."
        }
      ],
      "status": "final",
      "version": "1"
    }
  },
  "product_tree": {
    "full_product_names": [
      {
        "product_id": "CSAFPID-9080700",
        "name": "Product A"
      }
    ]
  },
  "vulnerabilities": [
    {
      "scores": [
        {
          "products": [
            "CSAFPID-9080700"
          ],
          "cvss_v3": {
            "version": "3.1",
            "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:C/C:H/I:H/A:H",
            "baseScore": 10,
            "baseSeverity": "CRITICAL"
          }
        },
        {
          "products": [
            "CSAFPID-9080700"
          ],
          "cvss_v3": {
            "version": "3.1",
            "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:C/C:H/I:H/A:H",
            "baseScore": 10,
            "baseSeverity": "CRITICAL"
          }
        }
      ]
    }
  ]
}
```

I'd interpret that the main focus of the test is on the CVSS-version per product. In this case, the code would be correct. If duplicate entries should be allowed (since there's at least no contradiction or so), it would make sense to check the vector string as well.

Reply:
> No, I was wondering how it's supposed to behave in such a case (modified the resource from your link a bit):
> 
> ```json
> {
>   "document": {
>     "category": "csaf_base",
>     "csaf_version": "2.0",
>     "publisher": {
>       "category": "other",
>       "name": "OASIS CSAF TC",
>       "namespace": "https://csaf.io"
>     },
>     "title": "Mandatory test: Multiple Scores with same Version per Product (failing example 1)",
>     "tracking": {
>       "current_release_date": "2021-07-21T10:00:00.000Z",
>       "id": "OASIS_CSAF_TC-CSAF_2.0-2021-6-1-07-01",
>       "initial_release_date": "2021-07-21T10:00:00.000Z",
>       "revision_history": [
>         {
>           "date": "2021-07-21T10:00:00.000Z",
>           "number": "1",
>           "summary": "Initial version."
>         }
>       ],
>       "status": "final",
>       "version": "1"
>     }
>   },
>   "product_tree": {
>     "full_product_names": [
>       {
>         "product_id": "CSAFPID-9080700",
>         "name": "Product A"
>       }
>     ]
>   },
>   "vulnerabilities": [
>     {
>       "scores": [
>         {
>           "products": [
>             "CSAFPID-9080700"
>           ],
>           "cvss_v3": {
>             "version": "3.1",
>             "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:C/C:H/I:H/A:H",
>             "baseScore": 10,
>             "baseSeverity": "CRITICAL"
>           }
>         },
>         {
>           "products": [
>             "CSAFPID-9080700"
>           ],
>           "cvss_v3": {
>             "version": "3.1",
>             "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:C/C:H/I:H/A:H",
>             "baseScore": 10,
>             "baseSeverity": "CRITICAL"
>           }
>         }
>       ]
>     }
>   ]
> }
> ```
> 
> I'd interpret that the main focus of the test is on the CVSS-version per product. In this case, the code would be correct. If duplicate entries should be allowed (since there's at least no contradiction or so), it would make sense to check the vector string as well.

@tschmidtb51 The standard seems to be vague in this regard. Currently our implementation, would say that this is an INCORRECT CSAF document, which in my opinion makes sense, since a duplicate doesn't make sense. Would that be ok?

Reply:
@oxisto I'm not fully convinced that the standard is vague at that point:

> For each item in `/vulnerabilities` it MUST be tested that the same Product ID is not member of more than one CVSS-Vectors with the same version.

Even if you check the `vector` value directly, you would still have 2 vectors with the same version (acknowledging though that the vector strings are the same).

So to answer your question: Yes, a CSAF with a duplicated CVSS item as shown above is invalid. 
(I just double-checked: Secvisogram does flag it correctly as invalid.)